types/convert.go

```diff
@@ -173,6 +173,48 @@ func ConvertFloatToUint(sc *stmtctx.StatementContext, fval float64, upperBound u
 	return uint64(val), nil
 }
 
+// ConvertDecimalToUint converts a decimal to an uint by string instead of float to avoid float overflow (#10181).
+func ConvertDecimalToUint(sc *stmtctx.StatementContext, d *MyDecimal, upperBound uint64, tp byte) (uint64, error) {
+	// since MyDecimal.ToString() doesn't use scientific notation now, it's safe to ignore it.
+	// TODO: handle scientific notation
+	str := string(d.ToString())
+	var intStr, fracStr string
+	p := strings.Index(str, ".")
+	if p == -1 {
+		intStr = str
+	} else {
+		intStr = str[:p]
+		fracStr = str[p+1:]
+	}
+	for len(intStr) > 0 && intStr[0] == '0' {
+		intStr = intStr[1:]
+	}
+	if sc.ShouldClipToZero() && intStr[0] == '-' {
+		return 0, overflow(str, tp)
+	}
+
+	var round uint64
+	if fracStr != "" {
+		tmp := "5" + strings.Repeat("0", len(fracStr)-1)
+		if fracStr >= tmp {
+			round++
+		}
+	}
+
+	upperBound -= round
```

Comment:
What does line204~208 mean?

Reply:
Check if this value is overflow by comparing string directly to avoid uint overflow when converting str to uint.